libraries/platforms/dromajo-vcs/software/src/argp/argp-parse.c

```diff
@@ -0,0 +1,992 @@
+    if (parser->state.next >= parser->state.argc
+      || (parser->state.flags & ARGP_NO_ARGS))
+    /* Indicate that we're done.  */
+    {
+      *arg_ebadkey = 1;
+      return EBADKEY;
+    }
+    else
+    /* A non-option arg; simulate what getopt might have done.  */
+    {
+      opt = KEY_ARG;
+      optarg = parser->state.argv[parser->state.next++];
+    }
+  }
+
+  if (opt == KEY_ARG)
+    /* A non-option argument; try each parser in turn.  */
+    err = parser_parse_arg (parser, optarg);
+  else
+    err = parser_parse_opt (parser, opt, optarg);
+
+  if (err == EBADKEY)
+    *arg_ebadkey = (opt == KEY_END || opt == KEY_ARG);
+
+  return err;
+}
+
+/* Parse the options strings in ARGC & ARGV according to the argp in ARGP.
+   FLAGS is one of the ARGP_ flags above.  If END_INDEX is non-NULL, the
+   index in ARGV of the first unparsed option is returned in it.  If an
+   unknown option is present, EINVAL is returned; if some parser routine
+   returned a non-zero value, it is returned; otherwise 0 is returned.  */
+error_t
+__argp_parse (const struct argp *argp, int argc, char **argv, unsigned flags,
+        int *end_index, void *input)
+{
+  error_t err;
+  struct parser parser;
+
+  /* If true, then err == EBADKEY is a result of a non-option argument failing
+     to be parsed (which in some cases isn't actually an error).  */
+  int arg_ebadkey = 0;
+
+  // Sripathi:
+  // Forward declare the struct pointers since malloc is being used instead of alloca
+  struct argp_child *child;
+  struct argp *top_argp;
+  if (! (flags & ARGP_NO_HELP))
+  /* Add our own options.  */
+  {
+    // Sripathi:
+    // Originally alloca was used but since newlib does not implement that
+    // we are switching over to malloc which is a lot slower
+    child = (struct argp_child *) malloc (4 * sizeof (struct argp_child));
+    top_argp = (struct argp *) malloc (sizeof (struct argp));
+
+    /* TOP_ARGP has no options, it just serves to group the user & default
+    argps.  */
+    memset (top_argp, 0, sizeof (*top_argp));
+    top_argp->children = child;
+
+    memset (child, 0, 4 * sizeof (struct argp_child));
+
+    if (argp)
+      (child++)->argp = argp;
+    (child++)->argp = &argp_default_argp;
+    if (argp_program_version || argp_program_version_hook)
+      (child++)->argp = &argp_version_argp;
+    child->argp = 0;
+
+    argp = top_argp;
+  }
+
+  /* Construct a parser for these arguments.  */
+  err = parser_init (&parser, argp, argc, argv, flags, input);
+
+  if (! err)
+  /* Parse! */
+  {
+    while (! err)
+      err = parser_parse_next (&parser, &arg_ebadkey);
+    err = parser_finalize (&parser, err, arg_ebadkey, end_index);
+  }
+
+  // Sripathi:
+  // malloc was used to allocate space for the argp_child and argp structs
+  // alloca automatically freed the memory when the function returned
+  // but since we are using malloc, we need to explicitly free it to
+  // avoid memory leaks
+  free(child);
+  free(top_argp);
```

Comment:
I have some most likely erroneous code/implementation here and I wanna know if this was a fine replacement for alloca() or should I actually implement it in assembly or something. I have a feeling this might be a possible source of the heisenbug because I am playing around with memory (fire)